### PR TITLE
HW Supported Channels for channel pool validation

### LINKED ISSUE #961: [BUG] agent: MAP-4.5.1_ETH fails with "No Channel Preference Report message found"
Job: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/468547961

logs: https://gitlab.com/prpl-foundation/prplMesh/-/jobs/468547961/artifacts/browse/logs/FAIL/MAP-4.5.1_ETH/

### PR DESCRIPTION
**Pull request by itayx**
currently channel pool validated by static table supported_channels_array[] of channels in db::set_dcs_channel_pool()

This is  incorrect  since radio setup could vary with different country configuration.

We need permanent prplmesh solution:

create  new structure  in node.h     std::vector<beerocks_message::sWifiChannel> phy_supported_channels
save GET_HW_FEATURES report during in slave_join to DB.
see ap_manager_thread::handle_hostapd_attached

ap_wlan_hal->read_supported_channels();

use it to Validate new dcs channel pool in db::set_dcs_channel_pool()
remove static table supported_channels_array[]  from db::set_dcs_channel_pool()

Task #686 
Fixes #961

### CODE REVIEW

File: common/beerocks/bwl/common/base_wlan_hal.cpp [hunk 1]
```diff
@@ -32,6 +32,7 @@ base_wlan_hal::base_wlan_hal(HALType type, std::string iface_name, IfaceType ifa
 
     // Initialize complex containers of the radio_info structure
     m_radio_info.supported_channels.resize(128 /* TODO: Get real value */);
+    m_hw_supported_channels.resize(128 /* TODO: Get real value */);
```
Comment (kantera800) [inline common/beerocks/bwl/common/base_wlan_hal.cpp:35]: why copy paste,   
find the max available channels , use the 5G (it has more channels) 

edited:
why not use also SUPPORTED_CHANNELS_LENGTH

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp [hunk 2]
```diff
@@ -1569,9 +1569,14 @@ bool ap_wlan_hal_dwpal::read_supported_channels()
             return false;
         }
 
-        // LOG(DEBUG) << "numOfValidArgs[0]= " << numOfValidArgs[0] << " chan= " << chan;
-        // LOG(DEBUG) << "numOfValidArgs[1]= " << numOfValidArgs[1] << " dfs_state= " << dfs_state;
-
+        LOG(DEBUG) << "numOfValidArgs[0]= " << numOfValidArgs[0] << " chan= " << chan;
```
Comment (kantera800) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1572]: please do cleanups in separate commit

**edited:**
 leave there prints , its for debugging.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp [hunk 3]
```diff
@@ -1569,9 +1569,14 @@ bool ap_wlan_hal_dwpal::read_supported_channels()
...
+        LOG(DEBUG) << "numOfValidArgs[1]= " << numOfValidArgs[1] << " dfs_state= " << dfs_state;
+        
+        m_hw_supported_channels[i].bandwidth   = 20;
+        m_hw_supported_channels[i].channel     = chan;
+        m_hw_supported_channels[i].bss_overlap = 0;
```
Comment (kantera800) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1577]: I'm pretty sure we don't need or use "bandwidth" and "bss_overlap" in the m_hw_supported_channels structure

  GET_HW_FEATURES reply from hostapd gives us:
Allowed channel: mode=2 chan=144 freq=5720 MHz max_tx_power=24 dBm (DFS state = available)

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 4]
```diff
@@ -1508,6 +1518,8 @@ void ap_manager_thread::handle_hostapd_attached()
 
     // Copy the channels supported by the AP
     copy_radio_supported_channels(ap_wlan_hal, notification->params().supported_channels);
+    // Copy the channels supported by the AP
+    copy_radio_supported_channels(ap_wlan_hal, notification->params().hw_supported_channels, true);
```
Comment (kantera800) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:1522]: this is not correct , you shouldn't **over write** supported channels with hw_supported channels.
For now hw_supported channels should be copied  along with the supported_channels in the  SLAVE_JOIN message 

in the future supported_channels will be removed.

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 5]
```diff
@@ -34,9 +34,12 @@ using namespace beerocks::net;
 //////////////////////////////////////////////////////////////////////////////
 
 static void copy_radio_supported_channels(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
-                                          beerocks::message::sWifiChannel supported_channels[])
+                                          beerocks::message::sWifiChannel supported_channels[],
+                                          bool use_hw_supported_channels = false)
```
Comment (kantera800) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:38]: you shouldn't copy hw_supported_channels to supported_channels.
please remove this commit.

File: controller/src/beerocks/master/db/db.cpp [hunk 6]
```diff
@@ -1165,7 +1165,7 @@ int db::get_hostap_conducted_power(std::string mac)
 }
 
 bool db::set_hostap_supported_channels(std::string mac, beerocks::message::sWifiChannel *channels,
-                                       int length)
+                                       int length, bool use_hw_supported_channels)
```
Comment (kantera800) [inline controller/src/beerocks/master/db/db.cpp:1168]: again something isn't right with your solution, 
this flag is not needed (same for the other functions)

edit:
please use dedicated  method for it 
`set_hw_supported_channels()`

**Review by kantera800 — CHANGES_REQUESTED** [review]
the solution seems a bit off track.

**hw_supported_channels** holds a list of  channels that FW supports.
**supported_channels** holds a list of current radio supported channels.
so in case of static channel,  supported_channels will hold on the only channel it supports.

hw_supported_channels **shouldn't overwrite** AP's current supported_channels list.

please make the appropriate changes.

edit:
I can see now that there is no overwrite, but PR still needs work.

File: agent/src/beerocks/slave/ap_manager_thread.cpp (hunk 4 above)
Comment (itayx) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:1522]: I don't overwrite the supported_channels
If you look at line 1522, it functions like so:
copy the supported channel from `ap_wlan_hal` into `hw_supported_channels` and use the HW supported channels in the `bwl`

File: agent/src/beerocks/slave/ap_manager_thread.cpp (hunk 5 above)
Comment (itayx) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:38]: like I said before, I don't copy the `hw_supported_channels` into `supported_channels`
since when called I use the notification params' **hw_supported_channels**

Comment (kantera800) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:38]: In that case don't change this method.
just create a dedicated method for copying hw supported channels
`copy_hw_supported_channels()`

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 7]
```diff
@@ -67,10 +70,15 @@ static int8_t get_tx_power(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal)
 }
 
 static std::string
-get_radio_supported_channels_string(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal)
+get_radio_supported_channels_string(std::shared_ptr<bwl::ap_wlan_hal> &ap_wlan_hal,
```
Comment (kantera800) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:73]: also here , please create a dedicated method
`get_hw_supported_channels_string()`

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 8]
```diff
@@ -1526,6 +1530,8 @@ void ap_manager_thread::handle_hostapd_attached()
     LOG(INFO) << " current bw = " << ap_wlan_hal->get_radio_info().bandwidth;
     LOG(INFO) << " supported_channels = " << std::endl
               << get_radio_supported_channels_string(ap_wlan_hal);
+    LOG(INFO) << " hw_supported_channels = " << std::endl
+              << get_radio_supported_channels_string(ap_wlan_hal, true);
```
Comment (kantera800) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:1534]:
```suggestion
              << get_hw_supported_channels_string(ap_wlan_hal);
```

File: agent/src/beerocks/slave/ap_manager_thread.cpp (hunk 4 above)
Comment (kantera800) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:1522]: then its very confusing.
If you want to copy hw_supported_channels to reply message then do it right after you read it from hostapd.
line 1473
`ap_wlan_hal->read_supported_channels();`

also, don't reuse `copy_radio_supported_channels()` by adding a flag, use dedicated function.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml [hunk 9]
```diff
@@ -271,6 +271,9 @@ sNodeHostap:
   supported_channels:
     _type: beerocks::message::sWifiChannel 
     _length: [ "beerocks::message::SUPPORTED_CHANNELS_LENGTH" ]          
+  hw_supported_channels:
```
Comment (tomereli) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:274]: Wait - what is the difference between `supported_channels` and `hw_supported_channels`?
The commit message should explain what are the AP supported channels and what are the HW supported channels.

File: common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h [hunk 10]
```diff
@@ -371,17 +371,24 @@ typedef struct sNodeHostap {
     uint8_t conducted_power;
     char driver_version[beerocks::message::WIFI_DRIVER_VER_LENGTH];
     beerocks::message::sWifiChannel supported_channels[beerocks::message::SUPPORTED_CHANNELS_LENGTH];
+    beerocks::message::sWifiChannel hw_supported_channels[beerocks::message::SUPPORTED_CHANNELS_LENGTH];
```
Comment (tomereli) [inline common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h:374]: Can be squashed to the previous commit.

**Review by tomereli — COMMENTED** [review]
I can't really review this before I understand what are the AP supported channels and what are the HW supported channels. This should be explained in the PR and in the commits which add the HW supported channels.

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 11]
```diff
@@ -1483,8 +1487,6 @@ void ap_manager_thread::handle_hostapd_attached()
 
             usleep(ACS_READ_SLEEP_USC);
         }
-    } else {
-        ap_wlan_hal->read_supported_channels();
```
Comment (kantera800) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:1479]: commit message should mention such changes.
something like:

> 
> We want always to save GET_HW_FEATURES when hostapd attached.
read_supported_channels() reads from GET_HW_FEATURES channel report and save supported channel list .
> But in case ACS report exists then its used instead of reading GET_HW_FEATURES .
> Since supported channels container is cleaned before reading the acs report read_supported_channels() can be read out side the **acs_enabled** condition.

File: common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp [hunk 12]
```diff
@@ -1550,7 +1550,7 @@ bool ap_wlan_hal_dwpal::read_supported_channels()
     FieldsToParse fieldsToParse[] = {
         {(void *)&chan, &numOfValidArgs[0], DWPAL_INT_PARAM, "chan=", 0},
         {(void *)dfs_state, &numOfValidArgs[1], DWPAL_STR_PARAM,
-         "(DFS state = ", sizeof(dfs_state)},
+        "(DFS state = ", sizeof(dfs_state)},
```
Comment (arnout) [inline common/beerocks/bwl/dwpal/ap_wlan_hal_dwpal.cpp:1553]: This change makes clang-format fail.

File: controller/src/beerocks/master/db/db.cpp [hunk 13]
```diff
@@ -1184,7 +1184,7 @@ bool db::set_hostap_supported_channels(std::string mac, beerocks::message::sWifi
         LOG(ERROR) << "No supported channels";
         return false;
     }
-
+
```
Comment (arnout) [inline controller/src/beerocks/master/db/db.cpp:1187]: Another clang-format error is introduced here.

**Review by arnout — CHANGES_REQUESTED** [review]
I can't say I'm happy with this PR...

First of all I think the naming is off. AFAIU, what is called "hw_supported_channels" here, really is "supported_channels", and what is called "supported_channels" is actually "channel_preferences". I think a few of @kantera800's comments are triggered by this confusion.

Second, I don't think it's right to use the sWifiChannel structure for something which is just a channel number, not an actual channel (with bandwidth).

Most importantly, though, it is yet again going in the wrong direction, using a VS message instead of the EasyMesh message. In this case, the EasyMesh way is already implemented - with AP Radio Basic Capabilities. So the controller should just parse that and save it in the database, rather than adding an extra set of channels to the VS message. Note that this will also cover my second general remark.

I can be convinced to relax on these general worries, but for now it's changes requested.

**Comment by kantera800** [comment]
> I can't say I'm happy with this PR...
> 
> First of all I think the naming is off. AFAIU, what is called "hw_supported_channels" here, really is "supported_channels", and what is called "supported_channels" is actually "channel_preferences". I think a few of @kantera800's comments are triggered by this confusion.
> 
> Second, I don't think it's right to use the sWifiChannel structure for something which is just a channel number, not an actual channel (with bandwidth).
> 
> Most importantly, though, it is yet again going in the wrong direction, using a VS message instead of the EasyMesh message. In this case, the EasyMesh way is already implemented - with AP Radio Basic Capabilities. So the controller should just parse that and save it in the database, rather than adding an extra set of channels to the VS message. Note that this will also cover my second general remark.
> 
> I can be convinced to relax on these general worries, but for now it's changes requested.

@arnout there is no confusion , I just wasn't sure if easymesh protocol already covers it so we went on the VS implementation to get the job done.
Now that you mention it easymesh way is better solution.

I checked the spec and current master implementation this is what i found:
Currently we build tlvApRadioBasicCapabilities and send it at slave_join.
but, the operation class list is generated still from the **supported_channels** list .
AFAIU in case of static channel we will have only 1 operating class in the tlvApRadioBasicCapabilities as a result  we could scan only 1 channel from the DCS channel pool.
So i see the following changes that has to be done:
1.  rename hw_supported_channels to supported_channels
2.  rename supported_channels to channel_preferences
3.  in add_ap_radio_basic_capabilities,  build  operation class list from **supported_channels**
4.  in controller, validate DCS channel pool against supported_channel list from previously saved tlvApRadioBasicCapabilities  data.

is this what you had in mind ?

edit:
@tomereli Just saw that in handle_cmdu_1905_autoconfiguration_WSC we are currently validating bss_info with radio_basic_caps so, AFAIU this change is needed unless we wouldn't pass certification when the radio operate with static channel.

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 14]
```diff
@@ -1469,9 +1469,13 @@ void ap_manager_thread::handle_hostapd_attached()
...
     LOG(DEBUG) << "handling enabled hostapd";
 
+    // Read HW supported channels by default.
+    ap_wlan_hal->read_supported_channels();
+    
     if (acs_enabled) {
         LOG(DEBUG) << "retrieving ACS report";
         int read_acs_attempt = 0;
+        // During each attempt the supported_channels are wiped and reset.
         while (!ap_wlan_hal->read_acs_report()) {
```
Comment (tomereli) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:1479]: I agree with @arnout - The ACS report provides the channel preference. The supported channels are the hw supported channels. This name change is critical.

**Comment by arnout** [comment]
> So i see the following changes that has to be done:
> 
>     1. rename hw_supported_channels to supported_channels
>     2. rename supported_channels to channel_preferences

I would swap those two :-)

>     3. in add_ap_radio_basic_capabilities,  build  operation class list from **supported_channels**
>     4. in controller, validate DCS channel pool against supported_channel list from previously saved tlvApRadioBasicCapabilities  data.
> 
> is this what you had in mind ?

It's exactly what I had in mind.

There's one more thing to take into account: the supported channels list should contain all supported bandwidths, so e.g. channel 52 is there twice, once for 20MHz and once for 40MHz. Also remember to take into account that the 80 and 160MHz bandwidths use center frequency, not base frequency. I *think* both are OK, but think about it as well.

**Comment by kantera800** [comment]
> > So i see the following changes that has to be done:
> > ```
> > 1. rename hw_supported_channels to supported_channels
> > 2. rename supported_channels to channel_preferences
> > ```
> 
> I would swap those two :-)
👍 
> 
> There's one more thing to take into account: the supported channels list should contain all supported bandwidths, so e.g. channel 52 is there twice, once for 20MHz and once for 40MHz. Also remember to take into account that the 80 and 160MHz bandwidths use center frequency, not base frequency. I _think_ both are OK, but think about it as well.
👍 

@arnout one thing i still don't understand.
 after this change,  how the controller will be notified of radio's static channel , is there a diffrent TLV for that?

File: agent/src/beerocks/slave/ap_manager_thread.cpp (hunk 4 above)
Comment (tomereli) [reply agent/src/beerocks/slave/ap_manager_thread.cpp:1522]:
```suggestion
    copy_supported_channels(ap_wlan_hal, notification->params().supported_channels);
```

File: agent/src/beerocks/slave/ap_manager_thread.cpp [hunk 15]
```diff
@@ -1516,6 +1518,8 @@ void ap_manager_thread::handle_hostapd_attached()
 
     // Copy the channels supported by the AP
     copy_radio_supported_channels(ap_wlan_hal, notification->params().supported_channels);
```
Comment (tomereli) [inline agent/src/beerocks/slave/ap_manager_thread.cpp:1520]:
```suggestion
    copy_channel_preference(ap_wlan_hal, notification->params().channel_preference);
```

File: controller/src/beerocks/master/db/db.cpp (hunk 6 above)
Comment (tomereli) [reply controller/src/beerocks/master/db/db.cpp:1168]: Please align to the new terminology - `set_hostap_supported_channels` and `set_hostap_channel_preference` - split this to 2 functions.

File: controller/src/beerocks/master/db/db.cpp [hunk 16]
```diff
@@ -2276,16 +2276,13 @@ bool db::set_channel_scan_pool(const sMacAddr &mac, const std::unordered_set<uin
...
-    static size_t supported_channels_array_size =
-        sizeof(supported_channels_array) / sizeof(uint8_t);
-    std::set<uint8_t> supported_channels(supported_channels_array,
-                                         supported_channels_array + supported_channels_array_size);
+    auto supported_channels =
+        get_hostap_supported_channels(network_utils::mac_to_string(mac), true);
     for (auto channel : channel_pool) {
-        if (supported_channels.find(channel) == supported_channels.end()) {
+        if (std::find_if(supported_channels.begin(), supported_channels.end(),
+                         [&channel](const beerocks::message::sWifiChannel &wifi_channel) {
```
Comment (tomereli) [inline controller/src/beerocks/master/db/db.cpp:2283]:
```suggestion
                         [&](const beerocks::message::sWifiChannel &wifi_channel) {
```

File: controller/src/beerocks/master/db/db.cpp [hunk 17]
```diff
@@ -2276,16 +2276,13 @@ bool db::set_channel_scan_pool(const sMacAddr &mac, const std::unordered_set<uin
...
-        64, 100, 104, 108, 112, 116, 120, 124, 128, 132, 136, 140, 144, 149, 153, 157, 161, 165};
-    static size_t supported_channels_array_size =
-        sizeof(supported_channels_array) / sizeof(uint8_t);
-    std::set<uint8_t> supported_channels(supported_channels_array,
-                                         supported_channels_array + supported_channels_array_size);
+    auto supported_channels =
+        get_hostap_supported_channels(network_utils::mac_to_string(mac), true);
     for (auto channel : channel_pool) {
-        if (supported_channels.find(channel) == supported_channels.end()) {
+        if (std::find_if(supported_channels.begin(), supported_channels.end(),
```
Comment (tomereli) [inline controller/src/beerocks/master/db/db.cpp:2282]:
```suggestion
        auto channel = std::find_if(supported_channels.begin(), supported_channels.end(),...);
        if (channel == supported_channels.end()) {...}
```

**Review by tomereli — CHANGES_REQUESTED** [review]
Please refactor this PR to use supported channels and channel preference as it should, then ask for a re-review.

**Comment by tomereli** [comment]
First - I fully agree with @arnout about the direction.

Regarding
> edit:
> @tomereli Just saw that in handle_cmdu_1905_autoconfiguration_WSC we are currently validating bss_info with radio_basic_caps so, AFAIU this change is needed unless we wouldn't pass certification when the radio operate with static channel.

Can you explain which test will fail due to this? Not sure I understand.

**Comment by kantera800** [comment]
> First - I fully agree with @arnout about the direction.
> 
> Regarding
> 
> > edit:
> > @tomereli Just saw that in handle_cmdu_1905_autoconfiguration_WSC we are currently validating bss_info with radio_basic_caps so, AFAIU this change is needed unless we wouldn't pass certification when the radio operate with static channel.
> 
> Can you explain which test will fail due to this? Not sure I understand.
>
AFAIU in case of static channel, ACS report overwrites **supported channels** with just this channel.
So, I assume that all tests that during handle_cmdu_1905_autoconfiguration_WSC rely on radio_basic_caps to include all radio **supported channels**  will go to teardown.

**Comment by tomereli** [comment]
> > First - I fully agree with @arnout about the direction.
> > Regarding
> > > edit:
> > > @tomereli Just saw that in handle_cmdu_1905_autoconfiguration_WSC we are currently validating bss_info with radio_basic_caps so, AFAIU this change is needed unless we wouldn't pass certification when the radio operate with static channel.
> > 
> > 
> > Can you explain which test will fail due to this? Not sure I understand.
> https://assets.grammarly.com/emoji/v1/1f91d.svg
> AFAIU in case of static channel, ACS report overwrites **supported channels** with just this channel.
> So, I assume that all tests that during handle_cmdu_1905_autoconfiguration_WSC rely on radio_basic_caps to include all radio **supported channels** will go to teardown.

You are correct. It is inherently wrong to use the ACS report for getting the supported channels, since I expect that if static channel is configured, the ACS report to be **empty** - I just verified this in the RAX40. @rmelotte this probably explains #961 -
In the CI, the DUT is rarely rebooted. The UCI default configuration is ACS, so `hostapd_cli -iwlan0 acs_report` will get some channel preference (not supported, but preferred).
Each time there is a scan, the ACS report gets updated. If there is no scan, the previous ACS report is fetched, but after 24 hours (according to one of my colleagues) it will be deleted, and the report will come out empty. This still has to be confirmed by the WLAN team in Intel, but a simple check on RAX40 in the CI will prove this check if you get acs_report, if its empty, reboot and check again. Wait 24 hours, then check again - should be empty again.

Bottom line is that I think this PR will solve #961 and possibly many others.

**Comment by arnout** [comment]
> after this change, how the controller will be notified of radio's static channel , is there a diffrent TLV for that?

I'm not sure what you mean with "static channel".

If it is the channel that it's currently operating at: after a channel selection request, the agent is supposed to send an Operating Channel Report CMDU to the controller. It should also do that when it changes channels on its own initiative.

It would actually make sense to send an unsollicited Operating Channel Report CMDU immediately after autoconfig search completes (i.e. just before sending M1). Otherwise, indeed, there's no way for the controller to know the current operating channel.

**Comment by kantera800** [comment]
> > after this change, how the controller will be notified of radio's static channel , is there a diffrent TLV for that?
> 
> I'm not sure what you mean with "static channel".

What i meant is that the radio **not** configured  with auto channel, so it can't switch channel on its own.
In this case the ACS report will have only one channel  (that's the "channel_preferences" we decided to rename earlier).
I want to make sure where exactly do we need to use channel_preferences (from acs) and where to use supported_channels (from radio capabilities /hw features) since this PR is going to change exactly that.
So actually my additional question is which easymesh message is responsible to notify the channel_preferences (from acs report)  to  the controller?

> 
> If it is the channel that it's currently operating at: after a channel selection request, the agent is supposed to send an Operating Channel Report CMDU to the controller. It should also do that when it changes channels on its own initiative.

that is when the radio configured with **auto channel** 
> 
> It would actually make sense to send an unsollicited Operating Channel Report CMDU immediately after autoconfig search completes (i.e. just before sending M1). Otherwise, indeed, there's no way for the controller to know the current operating channel.
>
 good point we need to verify that happens .

**Comment by arnout** [comment]
> 
> In this case the ACS report will have only one channel (that's the "channel_preferences" we decided to rename earlier).
> 

Oh, that will have to be checked... The Channel Preference TLV mentions only channels that are not preferred, with a preference value ranging from 0 (= this channel will not be used) to 14 (= just below maximum preference). Channels that were reported to be supported in the AP Radio Basic Capabilities but that are not mentioned in the Channel Preference TLV are considered to have maximum preference.

So, to be verified how the construction of the Channel Preference TLV treats an ACS report that has only a single channel. I suspect that it will create a Channel Preference TLV with all channels except the current one set to 0 (blocked). But that is not what we want - we probably should set the preference of the current channel to maximum, and the preference of all other channels to just below that (14, or maybe something like 12). Or at the very least, we should report no preference at all (so everything at maximum preference).

Ideally there should be a test of this :-)

> 
> I want to make sure where exactly do we need to use channel_preferences (from acs) and where to use supported_channels (from radio capabilities /hw features) since this PR is going to change exactly that.
> 

That's easy: where we use read_acs_report, it should be channel_preferences, and where we use read_supported_channels, it should be supported_channels.

> 
> So actually my additional question is which easymesh message is responsible to notify the > channel_preferences (from acs report) to the controller?
> 

Very difficult to find this: Channel Preference Report CMDU containing the Channel Preference TLV. And the supported_channels are sent in the AP Radio Basic Capabilities.

It would actually make sense to send an unsollicited Operating Channel Report CMDU immediately after autoconfig search completes (i.e. just before sending M1). Otherwise, indeed, there's no way for the controller to know the current operating channel.

> 
> good point we need to verify that happens .
> 

I can tell you: it doesn't happen at the moment.

As I mentioned earlier: the controller has no way to query the current channel, but now I think about it more, there is actually a trick that it can use. It can send a Channel Selection Request with no TLVs. The agent is supposed to send an Operating Channel Report message after receiving it, even if it doesn't switch channels. But of course, this is not tested in certification, so who knows how other agents will behave...

**Comment by kantera800** [comment]
> 
> But that is not what we want - we probably should set the preference of the current channel to maximum, and the preference of all other channels to just below that (14, or maybe something like 12). Or at the very least, we should report no preference at all (so everything at maximum preference).
> 

When I disable auto channel in my router I want him stay on the channel I selected ( bridging for example). 
AFAIU what you propose contradicts radio's "static channel" definition since now on the next BG scan DCS task will be able to switch to other channels. 

> 
> That's easy: where we use read_acs_report, it should be channel_preferences, and where we use read_supported_channels, it should be supported_channels.
> 

:) 
I was refering controller side. 
Currently supported channel (channel preference) used everywhere. 

p.s
i accidentally edited and reverted your last reply

**Comment by arnout** [comment]
> > But that is not what we want - we probably should set the preference of the current channel to maximum, and the preference of all other channels to just below that (14, or maybe something like 12). Or at the very least, we should report no preference at all (so everything at maximum preference).
> 
> When I disable auto channel in my router I want him stay on the channel I selected ( bridging for example).
> AFAIU what you propose contradicts radio's "static channel" definition since now on the next BG scan DCS task will be able to switch to other channels.

Right, so we really need 3 cases:
1. ACS -> channel_preference is set according to ACS scan.
2. No channel preference -> channel preference is 15 for current channel and 14 for everything else (to avoid switching channels if there's no reason to do so).
3. Static channel set by user -> channel preference is 15 for current channel and 0 or maybe 1 for all other channels.


> > That's easy: where we use read_acs_report, it should be channel_preferences, and where we use read_supported_channels, it should be supported_channels.
> 
> :)
> I was refering controller side.
> Currently supported channel (channel preference) used everywhere.

Controller side I don't think channel preferences are really used anywhere, except in the channel_selection task which has to be revamped anyway. But I haven't checked.

**Comment by tomereli** [comment]
Please note that the supported channels are available via NL80211 as implemented in https://github.com/prplfoundation/prplMesh/pull/977

**Comment by kantera800** [comment]
> Please note that the supported channels are available via NL80211as implemented in #977

we already use get_hw_features from hostapd.

But you reminded me of something else,
 since we are opening multiple sockets to nl (DWPAL, NL80211, arp monitor etc...) we are going to have a problem if we don't protect nl reply buffer from collision.
I already see it when scanning simultaneously on both radio's .   
not related to this PR.
